### ISSUE TITLE
Improve navigation and search in official documentation

### ISSUE THREAD

**Issue by ivandeex** [rclone/rclone#4784]
1. Please add search widget on the rclone site https://rclone.org
It can be as simple as a javascript wrapper around `https://google.com?site=rclone.org&q=search+phrase` (query syntax is sketched)

2. Some documentation pages are overgrown e.g. https://rclone.org/docs/. Finding a flag on large https://rclone.org/flags/ is hard. While navigation toolbar helps to find a page, scrolling in a large page is inconvenient. A contents block at the top would be nice.

**Comment by edwardxml** [comment]
> A contents block at the top

I agree with everything you have said. A toc contents card looks like the low hanging fruit
https://github.com/rclone/rclone/issues/4738

**Comment by ncw** [comment]
> 1. Please add search widget on the rclone site https://rclone.org
>    It can be as simple as a javascript wrapper around `https://google.com?site=rclone.org&q=search+phrase` (query syntax is sketched)

Is this allowed? When I've done this in the past it has involved getting API tokens and paying $$$. I guess that is for a results in the page search. If the search just opened a new tab with the google search in, I can't see Google complaining about that.

> 2\. Some documentation pages are overgrown e.g. https://rclone.org/docs/. Finding a flag on large https://rclone.org/flags/ is hard. While navigation toolbar helps to find a page, scrolling in a large page is inconvenient. A contents block at the top would be nice.

I made a prototype toc here https://test.rclone.org - it needs #4738 to be perfect

Feedback appreciated!

**Comment by edwardxml** [comment]
It is a long time since I used a variant of `https://google.com?site=rclone.org&q=search+phrase`. It worked but wasn't pretty. Hugo seems to have some options:

https://gohugo.io/tools/search/

**Comment by ivandeex** [comment]
@ncw @albertony 

Please try the search widget here:
https://rcdoc.deex.site/

Do you like it?
May I send a PR?

The html/js code is as simple as
```
<script>
function do_search() {
  var gs = document.google_search;
  gs.q.value = gs.words.value + " -site:forum.rclone.org";
  return true;
}</script>
<form action="https://google.com/search" id="google_search" name="google_search" target="_blank" onsubmit="do_search();">
  <input name="q" value="" type="hidden">
  <input name="words" value="" type="text">
  <input name="as_sitesearch" value="rclone.org" type="hidden">
  <button class="btn btn-primary" type="submit"><i class="fa fa-search search"></i></button>
</form>
```

I pulled the code from https://forum.rclone.org/search?q=junkee
and from https://forum.sia.tech/search?q=junkee

**Comment by ivandeex** [comment]
> Is this allowed? When I've done this in the past it has involved getting API tokens and paying $$$.

Using on-site google search forms is documented in O'Reilly. Googling for `as_sitesearch` will bring you a ton of howtos. Your forum engine has it embedded.
No tokens. No fees. No nothing.

**Comment by ivandeex** [comment]
Google also sells some junk called CSE API for money.
https://developers.google.com/custom-search/v1/overview
I guess we can live without it.

**Comment by albertony** [comment]
> Do you like it?

Yes, because it works! :)

Perhaps consider reducing the size of the search button to the right to match the textbox? I feel it is a bit... don't know how to explain it even, but perhaps a bit too dominant or something. Just a thought..

**Comment by ivandeex** [comment]
By default rclone.org site search will include results from forum.rclone.org and google does not provide a `query url` parameter to exclude it. I just append "-site:junk-feeder.org" to the search text.

**Comment by albertony** [comment]
> By default rclone.org site search will include results from forum.rclone.org and google does not provide a query url parameter to exclude it. I just append "-site:junk-feeder.org" to the search text.

I noticed that, but worked fine from what I could see.

**Comment by ivandeex** [comment]
> I made a prototype toc here https://test.rclone.org
> Feedback appreciated!

@ncw 

I can't see page TOCs. Your prototype is gone. :(
Please revive it

 > it needs #4738 to be perfect

AFAICS @edwardxml will not work on this in coming years.
Can we stick with imperfect layout?

**Comment by ivandeex** [comment]
> Perhaps consider reducing the size of the search button to the right to match the textbox? I feel it is a bit... don't know how to explain it even, but perhaps a bit too dominant or something. Just a thought..

Better now?
https://rcdoc.deex.site/?fight-cache=26

**Comment by albertony** [comment]
> Better now?

Yes, I think I would prefer that one.

**Comment by ivandeex** [comment]
> I made a prototype toc here https://test.rclone.org
> Feedback appreciated!

@ncw 

I tried https://test.rclone.org/docs/
Looks slightly strange when I hover over links in the Contents card.
I can't see with artist's eyes ;-). But I can use it and it works for me!

I vote :+1: to merge it!
I am 100% sure we'll see a patch request from a CSS professional less than one week after you publish it on rclone.org

> it needs #4738 to be perfect

AFAICS @edwardxml will not work on this in coming years.

**Comment by albertony** [comment]
> Looks slightly strange when I hover over links in the Contents card.

Cool effect! :)

Played around in browser devtools, and adding this stops the effect:

```
nav#TableOfContents li {
    overflow: hidden;
}
```

Then the existing "overflow: hidden" in `nav#TableOfContents ul` have no effect, and can be removed.

**Comment by ivandeex** [comment]
I redirected the nginx service at rcdoc.deex.site to test.rclone.org to see contents card and search widget togetger.
See https://rcdoc.deex.site/docs/

**Comment by ivandeex** [comment]
I also added a few of style overrides on https://rcdoc.deex.site/docs/
```
nav#TableOfContents ul {
    margin: 0.3rem;
}
nav#TableOfContents li {
    overflow: hidden;
    padding-top: 2px;
    white-space: normal;
}
```
@ncw 
If you like how the contents is looking now, you can tweak the code on _test.rclone.org_

**Comment by ivandeex** [comment]
Note: I'd recommend to replace `nav#TableOfContents` by more clean `#TableOfContents`.

**Comment by ivandeex** [comment]
I'd like to bring the page navigation in the docker online documentation as an example of well thought layout. You can try it on a wide-screen desktop or from smartphone. See how it handles deep'n'large content hierarchy. https://docs.docker.com/engine/reference/run/
We are far from that, yet this ticket is a baby step in the right direction.

**Comment by ncw** [comment]
> Please try the search widget here:
> https://rcdoc.deex.site/
> 
> Do you like it?
> May I send a PR?

Yes please!

> I tried https://test.rclone.org/docs/
> Looks slightly strange when I hover over links in the Contents card.

Yes that was an experiment so that the links use less vertical space. It also allows the hierarchy of links to be visible.

> I can't see with artist's eyes ;-). But I can use it and it works for me!

;-)

> I am 100% sure we'll see a patch request from a CSS professional less than one week after you publish it on rclone.org

Ha, ha! Yes you are probably right.

One thing that would help a lot is to make the page wider... I recently figured out how to do this in bootstrap.... This would let the contents column be wider which would make the layout a lot easier.

What do you think of this?  https://test.rclone.org ?

@ivandeex if you point me at the commit for the search, I'll merge that in too or add stuff to the `fix-toc` branch.

**Comment by ivandeex** [comment]
> @ivandeex if you point me at the commit for the search, I'll merge that in too or add stuff to the `fix-toc` branch.

#5411 

Note: i'm AFK till septmbr - can't code. still able to add minor comnents from top of head or press merge buttons on mobile github.

**Comment by ivandeex** [comment]
implemented since 1.56.0
closing
cc @ncw @albertony
